crates/sui-core/src/execution_cache/writeback_cache.rs

```diff
@@ -748,11 +748,19 @@ impl WritebackCache {
     fn revert_state_update_impl(&self, tx: &TransactionDigest) -> SuiResult {
         // TODO: remove revert_state_update_impl entirely, and simply drop all dirty
         // state when clear_state_end_of_epoch_impl is called.
-        let (_, outputs) = self
-            .dirty
-            .pending_transaction_writes
-            .remove(tx)
-            .expect("transaction must exist");
+        // Futher, once we do this, we can delay the insertion of the transaction into
+        // pending_consensus_transactions until after the transaction has executed.
+        let Some((_, outputs)) = self.dirty.pending_transaction_writes.remove(tx) else {
+            assert!(
+                !self.is_tx_already_executed(tx).expect("read cannot fail"),
```

Comment:
This will most likely hit DB given that we are doing a negative check? Should we make it debug_assert for that reason?

Reply:
Assuming its a recent tx, it may very well hit the cache due to #16968 

But in any case, it only runs at reconfig, so the perf hit should not matter very much, and this check failing would be a sign that something has gone very wrong.

Reply:
> Assuming its a recent tx, it may very well hit the cache due to https://github.com/MystenLabs/sui/pull/16968

But in this case, it's guaranteed to be not executed, hence it must not be in the cache?
Agree that since this only happens during reconfig, perf is probably not an issue

Reply:
I think you missed the `!` in the assert condition. This is asserting that if the tx is not in dirty, it is also not executed, since that case corresponds to the case where the tx was committed, and we should never revert a committed tx. So, if it is executed but not dirty, it could be in the cache.

the case where it hasn't been executed at all is totally allowable and expected